### DISCUSSION TITLE
Why does exec start a shell?

### DISCUSSION THREAD

**Discussion by h43z** [i3/i3#6310] [Q&A] [answered]
Just wondering why does exec start a shell and then in that shell the program? Why not just start it directly or have an option to start it directly.

**Comment by okraits** [comment]
https://i3wm.org/docs/userguide.html#exec

* A shell provides features like globbing, input/output redirection, logical operators like `&&`, etc.
* Some programs rely on specific variables etc. existing in the environment they are executed in. A shell provides that.

**Reply by h43z** [reply]
I get that there are advantages. But what if I don't need all of that. Isn't it kind of unnecessary and wasteful to have an extra process (sh) for every command I want to exec.

**Reply by WJCFerguson** [reply, marked as answer]
Would we not get the advantages if the command was called from the shell using `exec`, so the intermediate process would cease to exist?

**Reply by h43z** [reply]
Thank you. You are of course absolutely right I don't know how I wasn't able to see that.

**Reply by WJCFerguson** [reply]
Actually I wrote that kind of agreeing that i3 was doing something sub-optimal leaving all the shell processes as parents of i3 `exec`'d processes.   Seems to me `i3` should be executing the command with `exec` prepended, to get the functionality of executing through the shell, but without leaving a parent process hanging around for no reason.  

But in the meantime it does seem like if you tell i3 to `exec exec <command>` we get what we want (no pointless parent process).

**Reply by WJCFerguson** [reply]
yep - a bit of a spin around my config, making all `exec` become `exec exec`, gets rid of a mass of `sh -c` processes.  I also prepended an `exec` to my `i3blocks` config commands.

I'm very curious to know if there actually is a good reason that should not be the default behavior.  Do the parent `sh -c` processes provide anything?
